File: app/src/main/java/su/linka/pictures/TTS.java
package su.linka.pictures;

import android.content.Context;
import android.os.Build;
import android.speech.tts.TextToSpeech;

import java.io.File;
import java.io.IOException;
import java.util.UUID;

public class TTS {

    private final Context context;
    private final TextToSpeech tts;

    public TTS(Context context){
        this.context = context;
        tts = new TextToSpeech(context, new TextToSpeech.OnInitListener() {
            @Override
            public void onInit(int status) {

            }
        });
    }

    public void speak(String text){
        tts.speak((CharSequence) text, TextToSpeech.QUEUE_FLUSH, null, null);
    }

    public File speakToBuffer(String text) throws Exception {
        File file = new File(context.getCacheDir(), UUID.randomUUID().toString()+".wav");

            int res = tts.synthesizeToFile(text, null, file,null);
            if(res!=TextToSpeech.SUCCESS) throw new Exception("synth error");
        return file;

    }
}

File: app/src/main/java/su/linka/pictures/TTS.kt
package su.linka.pictures

import android.content.Context
import android.os.Build
import android.speech.tts.TextToSpeech
import java.io.File
import java.util.UUID

class TTS(context: Context) {

    private val appContext = context.applicationContext
    private val tts: TextToSpeech = TextToSpeech(appContext) { /* no-op */ }

    fun speak(text: String) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            tts.speak(text, TextToSpeech.QUEUE_FLUSH, null, UUID.randomUUID().toString())
        } else {
            @Suppress("DEPRECATION")
            tts.speak(text, TextToSpeech.QUEUE_FLUSH, null)
        }
    }

    @Throws(Exception::class)
    fun speakToBuffer(text: String): File {
        val file = File(appContext.cacheDir, "${UUID.randomUUID()}.wav")
        val result = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            tts.synthesizeToFile(text, null, file, UUID.randomUUID().toString())
        } else {
            @Suppress("DEPRECATION")
            tts.synthesizeToFile(text, null, file.absolutePath)
        }
        if (result != TextToSpeech.SUCCESS) {
            throw Exception("synth error")
        }
        return file
    }

    fun shutdown() {
        tts.shutdown()
    }
}
